src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs

```diff
@@ -564,11 +718,184 @@ public override ConstructorInfo[] GetConstructors(BindingFlags bindingAttr)
+            {
+                throw new ArgumentException(SR.Argument_MustBeInterface, nameof(interfaceType));
+            }
+
+            if (!IsInterfaceImplemented(interfaceType))
+            {
+                throw new ArgumentException(SR.Argument_InterfaceNotFound, nameof(interfaceType));
+            }
+
+            if (IsGenericParameter)
+            {
+                throw new InvalidOperationException(SR.Argument_GenericParameter);
+            }
+
+            // SZArrays implement the methods on IList`1, IEnumerable`1, and ICollection`1 with
+            // SZArrayHelper and some runtime magic. We don't have accurate interface maps for them.
+            if (IsSZArray && interfaceType.IsGenericType)
+            {
+                throw new ArgumentException(SR.Argument_ArrayGetInterfaceMap);
+            }
+        }
+
         [DynamicallyAccessedMembers(GetAllMembers)]
         public override MemberInfo[] GetMembers(BindingFlags bindingAttr) => throw new NotSupportedException();
-        public override bool IsAssignableFrom([NotNullWhen(true)] Type? c) => throw new NotSupportedException();
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+            Justification = "The GetInterfaces technically requires all interfaces to be preserved" +
+                "But in this case it acts only on TypeBuilder which is never trimmed (as it's runtime created).")]
+        public override bool IsAssignableFrom([NotNullWhen(true)] Type? c)
+        {
+            if (IsTypeEqual(this, c))
+            {
+                return true;
+            }
+
+            if (c is TypeBuilderImpl tb)
+            {
+                if (tb.IsSubclassOf(this))
+                {
+                    return true;
+                }
+
+                // in order to be a parent of interface this should be an interface
+                if (!IsInterface)
+                {
+                    return false;
+                }
+
+                Type[] interfaces = tb.GetInterfaces();
+                for (int i = 0; i < interfaces.Length; i++)
+                {
+                    // IsSubclassOf does not cover the case when they are the same type.
+                    if (IsTypeEqual(interfaces[i], this))
+                    {
+                        return true;
+                    }
+
+                    if (interfaces[i].IsSubclassOf(this))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        internal static bool IsTypeEqual(Type t1, Type? t2)
+        {
+            if (t1 is TypeBuilderImpl rtb1)
+            {
+                if (t2 is TypeBuilderImpl tb && ReferenceEquals(rtb1, tb))
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                return t1 == t2;
```

Comment:
Is there any reason not to check this first and only perform the more complex check if it is `false`?

Reply:
Good point